haskellPackages.hakyll: remove unnecessary patch, unbreak

erictapen opened the pull request:
###### Motivation

Unbreak `hakyll`.

###### Things done

Removed obsolete patch, as it's [replacement](https://github.com/jaspervdj/hakyll/pull/703) was merged and is in the current version.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---
@GrahamcOfBorg build haskellPackages.hakyll

erictapen committed:
haskellPackages.hakyll: remove unnecessary patch, unbreak

Infinisil commented:
I think you should instead remove `hakyll` (and dependents of it, if they build) from the list of broken packages here: https://github.com/NixOS/nixpkgs/blob/eca09e0d997546e005241d7c8cd6f8c999bf8496/pkgs/development/haskell-modules/configuration-hackage2nix.yaml#L5088

on the haskell-updates branch. Then the next automated update (probably by @peti) of `hackage-packages.nix` will not mark is at broken anymore. Remove the patch on that branch too.

erictapen commented:
Ok I changed the base to `haskell-updates` and removed everything from the `broken` section in `configuration-hackage2nix.yaml` that depends on `hakyll`. But I didn't do any builds, because that would be too time/bandwidth/storage consuming for me right now.

I find this workflow to be quite a frustrating one and I would like to explain here how I come to that conclusion.

First of all this is a commit that doesn't change anything to the better for me. If I build `hakyll`, I still get an error that the package is broken. I can't tell OfBorg to build it, and I can't easily prove to anybody else, that the packages builds now. For me I still need to remove the `broken` attribute with `overrideCabal` until the package set is regenerated.

Second, this leaves me with reverse dependency resolution, a task that I've never needed to do in other sections of Nixpkgs. Why should I care wether `agda-snippets-hakyll` builds? I fixed `hakyll` and every other problem with `agda-snippets-hakyll` should be subject to somebody trying to fix that package. Still in the current situation I need to remove `agda-snippets-hakyll` from `configuration-hackage2nix.yaml` because otherwise it could stay marked as broken while it actually could build.

Regarding this new practice of marking packages automatically as broken I agree with the reasoning @peti laid out [here](https://github.com/NixOS/nixpkgs/issues/59409#issuecomment-483187716):

> The change prevents people from wasting half an hour of their lives trying to compile a package that we already know will not succeed.

So I wonder wether it would be possible to only mark packages as broken if theyself fail to build? Packages with a broken dependency will fail to evaluate anyway, even if they are not marked as broken. This way I wouldn't need to work my way through the package set and find reverse dependencies.

Atm this blew up the time I take to simply remove a patch from ~30min to 3h. I'm already more of a low volume contributer but if this stays this way I'll find even less time to upstream my patches…

Still: Thanks for all the work.

Infinisil commented:
Yeah it is a bit cumbersome, I hope in the future only failing builds themselves will get marked as broken and not all dependents too.

Infinisil commented:
I pushed a commit to fix a dependency of hakyll on that branch.

The packages unmarked as broken often still don't build, but I think that's still fine.

peti commented on pkgs/development/haskell-modules/configuration-hackage2nix.yaml:
```diff
@@ -5085,27 +5084,6 @@ broken-packages:
   - hakismet
   - hakka
   - hako
-  - hakyll
```
Please don't unmark all those broken packages unless you are certain (i.e. you have personally verified) that they actually compile. I'm pretty sure that not all of those packages will build even if `hakyll` itself is fixed.

Infinisil replied:
Okay let's only mark hakyll as unbroken then. I have tested a couple others and yeah a lot of them still fail to build.

endgame replied:
Do broken packages get automatically added to this file? If so, then unmarking revdeps, having them fail and get re-added doesn't seem so bad.

peti commented:
> I wonder whether it would be possible to only mark packages as broken if theyself fail to build?

Attempts to build packages that depend on a broken packages result in an evaluation error on Hydra, unfortunately, so we cannot leave those builds enabled.

endgame commented:
I agree with most of what @erictapen wrote, and especially the "thanks for all your work" part. I tripped over broken `hakyll` last night, made a PR and was notified about this one overnight.

I disagree that it's impossible to prove it works: I can at least tell people that running `nix-build -E '(import ./. { config.allowBroken = true; }).whatever` builds. It does seem like a sticky trap for packages, though: they get marked broken easily but it's not so easy to get them unmarked. #61536 , which unmarks the lattices package, which has been stuck for a week.

What's needed to get either this or my hakyll PR ( #61875 ) over the line?

Infinisil commented:
Alright I merged your PR's @endgame. Note that you can also use `nix-build -A whatever --arg config '{ allowBroken = true; }'` :)

peti commented:
> Do broken packages get automatically added to this file?

No. That file is maintained manually.
